### ISSUE TITLE
User installation

### ISSUE THREAD

**Issue by jayvdb** [storyscript/community#8]
Underlying the problems of https://github.com/storyscript/community/issues/7 is that `pip` is a terrible tool to install/manage software, except in an isolated venv with nothing else in it.  Python devs know this, but other devs are not so familiar with the problems.  Most users do not want to create new venv for every piece of software, and switch between them whenever they want to run a different piece of software.  No `pip install --user foo` like we recommend at https://docs.storyscript.io/quick-start/#getting-started is not the answer - it only side-steps the root permissions problem (so the dev doesnt break their OS) and allows per-user site-packages (so two people dont have conflicts over the global site-packages).  `pip install --user storyscript` has a very high chance of breaking any other software the user has installed with `pip install --user ...` , or not installing correctly, especially at the moment as all projects are using `==` pins.  And that is before we get into the differences in pip versions which may exist on their system.

https://docs.storyscript.io/quick-start/#getting-started also lists brew which is of course a mature solution, and snap, which will help a small cohort of users.   AppImage and flatpaks would also be helpful, but like snap is going to only be used by a small set of potential users.

A large cohort of potential users are going to be choosing a conda package, or a chocolatey package (if isolated; i.e. not a chocolatey package which just uses the pip provider).

And of course linux users are going to choose a distro provided package if one exists.

A recent trend is to use tools like [`pip-run`](https://pypi.org/project/pip-run/) and [`pipx`](https://pypi.org/project/pipx/) when writing user facing documentation, so that the virtualenv stuff is done magically behind the scenes.  This makes the install more like `brew`.  I have used them on linux, but not on Windows.  `pip-run` has Windows CI, but very few users ([<60downloads/per](https://pypistats.org/packages/pip-run); `pipx` doesnt have Windows CI, and pipx has lots of raised issues about Windows bugs, however it has > 800 downloads/per day https://pypistats.org/packages/pipx .  A quick look at the Windows QA of pip-run suggests it wont hit the real problems with Windows -- but it should reduce regressions.

fades is quite good, but it depends on the cli being only a single .py file.  We might be able to achieve this by post-processing the cli module files into a single file - there are several tools which can do this.

There are a lot of other tools like these under [virtualenv manager @PyPI](https://pypi.org/search/?q=virtualenv+manager), [@GitHub](https://github.com/search?l=Python&o=desc&q=virtualenv+manager&s=stars&type=Repositories)

Another option is to use a Python .exe builder, especially for Windows (see Chocolatey above) but also for Windows.

**Comment by wilzbach** [comment]
> No pip install --user foo like we recommend at https://docs.storyscript.io/quick-start/#getting-started is not the answer - it only side-steps the root permissions problem (so the dev doesnt break their OS)

Agreed.

> https://docs.storyscript.io/quick-start/#getting-started also lists brew which is of course a mature solution, and snap, which will help a small cohort of users. AppImage and flatpaks would also be helpful, but like snap is going to only be used by a small set of potential users.

`snap` is a pretty neat way to ship things as it does auto-updates by default, but I agree that outside the Ubuntu world it's not very common yet.

> And of course linux users are going to choose a distro provided package if one exists.

The decision to only go with `brew` and `snap`  was done with the assumption that providing a special package for every distro has a poor work/return-value ratio.

> A large cohort of potential users are going to be choosing a conda package, or a chocolatey package (if isolated; i.e. not a chocolatey package which just uses the pip provider).

I'm not sure whether we should focus on providing `conda` or `chocolately` packages first as they don't reach everyone either.
Hence, imho, we should focus on these in decreasing priority for the story CLI:
1) Windows installer
2) Standalone archive or executable for Linux
3) Linux/macOS install script (can use the archive/executable from (2)), e.g.

```
curl -sSL https://storyscri.pt/install.py | python
```

**Comment by jayvdb** [comment]
I am very happy that we agree `pip install --user foo` is not acceptable.  I note that you didnt provide feedback regarding `pip-run` & `pipx` which are immediately adoptable solutions which eradicate most of the problems.

> The decision to only go with brew and snap was done with the assumption that providing a special package for every distro has a poor work/return-value ratio.

That is quite a vague basis for making decisions, and it sounds like you dont like the idea because word choice is exaggerating the ratio so that it justifies the existing decision.

Especially wrt "every distro", it is a bootstrap process.  Once it is done for Fedora (or openSUSE as an easier proxy for it) and Debian (much harder), and recently Arch, pretty much all distros will/can copy and paste and fiddle.  People using more obscure distros are likely to do that packaging if it hasnt been done yet for their distro, which helps to build our team.  And in all cases, distros have 'personal' repositories where users can build their own recipes before they are ready to get merged into the core distribution.  e.g. for copr Fedora and ppa for Debian/Ubuntu

Once the recipes are in the main collection, distro maintainers will maintain them, and actively assist the storyscript team when they identify problems -- and they will usually identify problems before the storyscript team does, as they run builds continuously and use different configurations.

The difficult part is not creating and maintaining those recipes for "Storyscript" software, as we can also get code changes merged in storyscript repos to facilitate this.  The hard part is creating the packages for dependencies, where you do not have rights to update their code to get it working correctly, and cant push out a release whenever you want.  That part has mostly been done.

It doesnt make sense to do have distro packages for software which is not yet stable, but that era of storyscript should be finishing soon and stability will soon become a key criteria users will consider when evaluating the software.

In addition to ease-of-install and stability, another very important benefit of packaging is licensing checks being done on all dependencies.  Often distros will tell software developers that the dependency they have added is a deal-breaker, and they need to get rid of it.  Developers don't tend to look at the licensing and maintainability of each dependency they add.

> I'm not sure whether we should focus on providing conda or chocolately packages first as they don't reach everyone either.

chocolatey is brew for Windows.  It is younger than brew, but there is no equivalent except enterprise solutions, and even they are now adding chocolatey support.  It doesnt reach everyone, but it the way to reach everyone on Windows in a scalable/maintainable way.

I wont bother replying wrt conda - it is like a distro that users will build their own recipes for if we don't.

> 1. Windows installer
> 2. Standalone archive or executable for Linux

So we agree these are important.  I see them as identical, and referred to them as "Python .exe builder".  The commonly used term is freezing, and a list of prominent options is at https://docs.python-guide.org/shipping/freezing/   Many support Windows and Linux and macOS, with  https://www.pyinstaller.org/ being the usual 'first' choice.

Another option that I believe is worth exploring is https://github.com/Nuitka ; it is just coming to the point of being usable for large apps (e.g. i.e. Django sized apps).

> Linux/macOS install script 
> curl -sSL https://storyscri.pt/install.py | python

This is very vague.  What will the script do?  Use pip?  Install a Python, create a dedicated venv, install the cli and create bin scripts?  Why not use/recommend pip-run/pipx?

**Comment by wilzbach** [comment]
> . I note that you didnt provide feedback regarding pip-run & pipx which are immediately adoptable solutions which eradicate most of the problems.

It doesn't because the user __needs_ to install the respective tools. Our target audience are people with limited programming experience, so every intermediate step is painful.

BTW regarding the two tools:

1) With pipx 0.14.0.0:

```
> pipx install story

Could not find package story. Is the name correct?
```

2) `pip-run` doesn't do permanent installations, only temporary ones.

> Especially wrt "every distro", it is a bootstrap process. Once it is done for Fedora (or openSUSE as an easier proxy for it) and Debian (much harder), and recently Arch, pretty much all distros will/can copy and paste and fiddle. People using more obscure distros are likely to do that packaging if it hasnt been done yet for their distro, which helps to build our team. 

Distros aren't an option, but it takes months or sometimes years (e.g. Debian) for an update to propagate and we do want new features or bug fixes to propagate quickly.

> Once the recipes are in the main collection, distro maintainers will maintain them, and actively assist the storyscript team when they identify problems -- and they will usually identify problems before the storyscript team does, as they run builds continuously and use different configurations.

That's a very optimistic way of seeing things. In my experience, it's really hard to find distro maintainers who are interested in just version bumping and they are often very inclined to drop packages if they start to break.

> It doesnt reach everyone, but it the way to reach everyone on Windows in a scalable/maintainable way.

I agree that chocolatey is nicer, but unfortunately I disagree that it reaches everyone. For comparison, the D Windows installer receives >3K downloads per day, the chocolately DMD package doesn't even provide the latest version and has 373 downloads as of now.

tl;dr: Installer >> chocolately (in terms of importance for Windows)

> Another option that I believe is worth exploring is https://github.com/Nuitka ; it is just coming to the point of being usable for large apps (e.g. i.e. Django sized apps).

Not sure whether it's production-ready yet, see e.g. https://www.reddit.com/r/Python/comments/8ha98n/nuitka_release_0530

> This is very vague. What will the script do? Use pip? Install a Python, create a dedicated venv, install the cli and create bin scripts? 

It's vague on purpose, because it's just a concept. People might not even have a recent version of Python installed on their machine, so an abstract installer would need to take care of this as well.
Anyhow, people really seem to like these one-line installation script that perform a local user installation without needing to worry about anything else.

- https://rustup.rs
- http://dlang.org/install
- https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py

It could even do a bit of system detection, e.g. if brew/snap available -> use these package managers.

> Why not use/recommend pip-run/pipx?

AFAICT `pip-run` doesn't solve the installation problem as it's just temporary and `pipx` doesn't seem to be mature yet. Though I do agree that recommending e.g.

```sh
pip install --user pipx
pipx install story
```

would be better than just displaying `pip install --user story`.

**Comment by jayvdb** [comment]
> Not sure whether it's production-ready yet, see e.g. reddit.com/r/Python/comments/8ha98n/nuitka_release_0530

That is from a year ago.  Posting links from stackoverflow and reddit, etc are not useful in a discussion unless they are also your own opinion, formed through experience.  Doing it suggests that you assume the other person doesnt know what they are talking about and that a random post by someone else on the internet is sufficient to discredit the other member of your community.  For my own part, I have been working with the core developer for months fixing corner cases. c.f. https://github.com/Nuitka/Nuitka/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+author%3Ajayvdb

I havent tried it on story cli yet, and wouldnt recommend attempting it until the next release is out, but that should be soon.

> AFAICT `pip-run` doesn't solve the installation problem as it's just temporary

A temporary installation is  isnt a problem.  For someone walking through steps in a getting-started, the fact it cant break their system can be a huge feature , and the overhead of reinstalling for each command is a minor performance problem.

The real problem with pip-run is IMO that it is an ugly user experience, both the first and subsequent use of the same software.

>> pipx install story
>Could not find package story. Is the name correct?

Hmm.  Any info on what type of environment you experience that in?
Could it be that you installed Python 2 pipx?  That gives me a different error

```
> python2 ~/.local/bin/pipx install story
Collecting story
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/14/32/41da9516cad929bc0340e45a2381b64e79055c1bd69274fa9c80301583a7/story-0.20.0.tar.gz (75kB)
    100% |████████████████████████████████| 81kB 1.4MB/s
Collecting click-completion==0.5.1 (from story)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/77/fd/2d7ec2b86cd4d487abf0b13dce58e98413096c45b9645470be0cb8de6ff2/click-completion-0.5.1.tar.gz
Collecting click-help-colors==0.5 (from story)
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/4e/88/10c3472e038bad3a504496df1cb657a8caf505b5cf4c6b95f3cae4cd4b10/click-help-colors-0.5.tar.gz
Collecting click-spinner==0.1.8 (from story)
  Cache entry deserialization failed, entry ignored
Requirement already satisfied: click==7.0 in /usr/lib/python2.7/site-packages (from story)
Collecting emoji==0.5.0 (from story)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/47/e7/cf76e0774b13459e61ffe16a042a35f42e292aeebba060b919c05b0854ab/emoji-0.5.0.tar.gz (41kB)
    100% |████████████████████████████████| 51kB 2.0MB/s
Collecting raven==6.9.0 (from story)
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading https://files.pythonhosted.org/packages/11/3a/b3e46b279b8bdd9eb55857d0e95044cad31732c80f628bb75e1e9e881a32/raven-6.9.0-py2.py3-none-any.whl (287kB)
    100% |████████████████████████████████| 296kB 1.1MB/s
Requirement already satisfied: requests~=2.20 in /usr/lib/python2.7/site-packages (from story)
Collecting storyscript==0.23.4 (from story)
  Cache entry deserialization failed, entry ignored
  Could not find a version that satisfies the requirement storyscript==0.23.4 (from story) (from versions: 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7, 0.0.8, 0.0.9, 0.0.10, 0.0.11, 0.0.12, 0.0.13, 0.0.14, 0.0.15, 0.0.16, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5, 0.1.6, 0.1.8, 0.1.9, 0.2.0, 0.2.1, 0.3.0, 0.3.1, 0.4.0, 0.4.1, 0.4.2, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.5.4, 0.6.0, 0.6.1, 0.6.2, 0.7.0, 0.7.1, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.9.0, 0.9.1, 0.9.2, 0.9.3, 0.9.4, 0.9.5, 0.9.6, 0.9.7, 0.9.8, 0.9.9, 0.9.10, 0.9.11, 0.9.12, 0.9.13, 0.9.14, 0.9.15, 0.10.0, 0.11.0, 0.12.0, 0.13.0, 0.14.0, 0.14.1, 0.14.2, 0.14.3, 0.15.0, 0.15.1, 0.16.0, 0.16.1, 0.16.2, 0.17.0, 0.17.1, 0.17.2, 0.17.3, 0.18.0)
No matching distribution found for storyscript==0.23.4 (from story)
You are using pip version 9.0.1, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```
(Isnt pip wonderful' though to be fair our setup.py should be improved to prevent that mess, or using pyproject.toml would also prevent that mess.)

Anyway, `pipx` works well for me on Linux, Python 3.

```sh
> pipx install story
  installed package story 0.20.0, Python 3.7.2
  These apps are now globally available
    - story
done! ✨ 🌟 ✨
> which story
/home/jayvdb/.local/bin/story
> head -1 $(which story)
#!/home/jayvdb/.local/pipx/venvs/story/bin/python
> ls /home/jayvdb/.local/pipx/venvs/story/bin/
activate  activate.csh  activate.fish  chardetect  pwiz.py  __pycache__  python  python3  raven  story  storyscript  xerox
```

Creating our own shell script to replicate that seems a bit odd.

Your example of get-pip.py is a great one - it started as a simple script, and grew slowly vendoring lots of stuff until they decided it was easier for it to be a frozen version of pip.  We should avoid the same mistake.

Storyscript is not like D and Rust, etc, for fairly obvious reasons.

**Comment by jayvdb** [comment]
https://gitlab.com/mkleehammer/kelvin looks like a good option for freezing on Windows only.

While performing the freeze and CD would be better on a Windows CI worker, https://github.com/paulfurley/python-windows-packager is a recipe for doing it on Linux using Wine, but it is a bit old and may not work with Python 3.6+ for Windows, which was a big shift in the Windows packaging of CPython.

https://github.com/FelixWolf/pylop also looks interesting, but not mature enough to consider - seems to be a subset of py2exe feature-set.

https://github.com/albertosottile/py2exe seems to be where the Python 3 version is being developed.

["pyinstaller+cx_freeze"](https://github.com/search?q=pyinstaller+cx_freeze+is%3Aissue&type=Issues) has some good tidbits.

https://github.com/nvaccess/nvda/issues/8375 is over a year old, but provides a nice overview comparison of the tools at the time.

**Comment by wilzbach** [comment]
> Any info on what type of environment you experience that in?

Arch Linux. Python 3.7.4.
I can definitely find out what went wrong here on my machine if we want to display `pipx` install instructions, but it's not a good sign for its maturity.

> https://gitlab.com/mkleehammer/kelvin looks like a good option for freezing on Windows only.

In my experience building things with Wine always leads to problems. For example, because you can't create static binaries because MS forbids you to redistribute their C runtime (except you build with their Visual Studio).

Apart from that I think the only way to know what tools works best for Storyscript is to actually give them a try and bundle the CLI with them.

Also, I think all the tools you linked create a single standalone executable. Do you plan to wrap this into an installer afterwards? I ask, because we might want to use sth. like https://github.com/takluyver/pynsist directly.

> Your example of get-pip.py is a great one - it started as a simple script, and grew slowly vendoring lots of stuff until they decided it was easier for it to be a frozen version of pip. We should avoid the same mistake.

It's not only Pip, there are many others, e.g.
- https://cli-assets.heroku.com/install.sh
- https://get.docker.com
- https://yarnpkg.com/install.sh 
- https://github.com/composer/getcomposer.org/blob/master/web/installer
...

Anyhow, I agree that that such a script might result in quite a bit of maintenance, but it avoids us to rely on Python for the "official" installation ways and thus doesn't lock us into the Python.

That being said I think Heroku's installation page is a great example of what installation paths we should offer to our users:

https://devcenter.heroku.com/articles/heroku-cli#download-and-install

**Comment by jayvdb** [comment]
Note that any decent Windows solution relies on Windows CI, and that is something that you wanted to take control of: https://github.com/storyscript/storyscript/issues/1037 .
We need that implemented asap on at least hub, storyscript and cli repos.

> I can definitely find out what went wrong here on my machine if we want to display pipx install instructions, but it's not a good sign for its maturity.

yes please.  It is more mature than the current fallback solution: `pip install --user story`.  And simpler than any of the other approaches being discussed.

> Arch Linux. Python 3.7.4.

Vendor supplied `pip` or custom; version?

TBH, I'm not particularly concerned if it doesnt work on Arch, due to our market positioning :P  I'm much more interested in whether it works on the top distros and of particular concern is Windows.

Unless someone knows of a better option which is simple and doesnt increase our cost of ownership, I believe it is worth investigating whether pipx works on Windows.  Should be fairly simple to establish whether that is a viable approach.  I wouldnt want to put `pipx` into the docs until that is done, for fear of biting off more than we can chew.

> Also, I think all the tools you linked create a single standalone executable. Do you plan to wrap this into an installer afterwards? I ask, because we might want to use sth. like https://github.com/takluyver/pynsist directly.

I think an MSI is unavoidable, and not terribly complicated, once we have something to put in it.  I havent seen pynsist.  If it works, great.  I havent done Windows installers for 10 years, so have no strong opinions on which tools to use.  I used WiX back then and it was already a huge leap forward from the prior tools available at the time.

It would be nice if we could provide a Portable App.  Nuitka can do this; some of the others probably can, but I have no experience with them.

If we are going to go beyond a CLI, and provide native Windows integration of any sort https://github.com/indygreg/PyOxidizer and its underlying https://github.com/indygreg/python-build-standalone are possibly a way to go.

Another important solution is also a Docker image of the cli.  It is wildly overkill, but it also very consumable and has zero chance of breaking the users environment, so will be appealing to a set of users who want to try risk-free.  Not a high priority, but also it is not a difficult solution to provide and maintain, so cost/benefit is in its favour.

> In my experience building things with Wine always leads to problems. For example, because you can't create static binaries because MS forbids you to redistribute their C runtime (except you build with their Visual Studio).

Worth noting I looked into IronPython's Python 3 effort, and it is still too immature to use, especially if we are targeting Python 3.6.  There are a few Java runtimes which look like they are viable for our needs, but I didnt get beyond initial viability.

(further aside, not especially helpful atm, I have fully tested PyPy 7 on linux - all tests of all dependencies are passing 'sufficiently', all the way down.  Only a few minor failures.)

**Comment by jayvdb** [comment]
On the topic of Windows, which is a bit of a sticking point, an interim decision on desired/supported versions of Windows would be a good starting point.

**Comment by jayvdb** [comment]
There has been private push back on why I am investigating Nuitka when PyInstaller appears to be an easy solution that 'just works' locally.

Now that I have the Windows builds emitting exes, I am mostly past any feasibility concerns wrt Nuitka, and other concerns should be considered, including maintainability.  A "it just works" exe from PyInstaller appears to win in the maintainability department.

There are still a few caveats wrt Nuitka, as the story cli test suite isnt operational atm, I only have ~25% of Nuitka builds on Windows running their test suites, and I havent created static builds yet.

As previously mentioned above https://github.com/nvaccess/nvda/issues/8375 is a nice overview , which ended up going ahead with the new py2exe over the alternatives.

https://pyoxidizer.readthedocs.io/en/latest/comparisons.html is also a good read.

Some additional considerations

*Verify-ability*

First and foremost, most freezers create a runtime version which is significantly different from the environment used in testing, and normal test routines can not be run on the runtime version.
As a result, only black box testing of the resulting CLI is possible.

story CLI has zero CLI tests IIRC.  Some of its test suite is high level and closely maps to the CLI interface, so those could be quite quickly adapted to become black box tests of the CLI.

However, the larger problem is that none of the Storyscript components have what are typically called integration tests that actually interface with real running services, and any mocks of a real running service are either non-existent or are hiding very well and are not mocking a realistic service.
For that, the mocks need to perform http traffic playback.

In light of this, a freeze solution that only allows CLI testing poses a large risk, and would require extensive human QA initially, and would require intensive human QA for each release until the quality of the test suites in all Storyscript dependencies are improved.

Most freezers, including Nuitka and PyInstaller, work around this by maintaining their own *basic* test apps which they run in their CI to maintain compatibility with major Python frameworks and libraries.

However they will update those test apps when breakages happen, and typically they immediately drop support for the old version of the library.
This de-support will be noted in the release notes, and possibly documented somewhere more readable.
However most are like PyInstaller which only test against one version of dependencies and have only a simple list like
https://github.com/pyinstaller/pyinstaller/wiki/Supported-Packages
When there are known incompatibilities, they do not fail with a helpful message when freezing an old unsupported version.  Note PyInstaller doesnt provide CI for click core, let alone click plugins, nor are commonly used Windows console pty & ansi color emulation libraries listed, such as `colorama`, `termcolor, `cleo`, or readline & pyreadline, etc.  Testing those is hard -- their own test suites are works of art.

Nuitka has package level "freezing" support as a primary use-case, allowing each Python source package to be replaced with a drop-in Nuitka compiled DSO.
Each Nuitka built package can be tested with all other components being either Nuitka compiled or original, allowing testing suites to operate normally.
Note there is one significant bug in Nuitka which is blocking use of `mock`'s autospec functionality.  https://github.com/Nuitka/Nuitka/issues/535
This is the only significant limitation encountered when running test suites of ~250 core Python libraries on Linux (only a few hours has been spent on performing Windows testing of dependencies; too early to reach conclusions).

A per-package verifiability process, running the package test suite for each package, reduces the risk of future release of dependencies causing unforeseen incompatibities with freezing.  The freeze might work, so CI "here" is green, but the frozen exe isnt being rigorously tested.

It might be possible to run the test suite on each package with PyInstaller's one-dir mode.  I did briefly attempt it but was ont successful in getting test suites working en-mass.
c.f. https://build.opensuse.org/request/show/732148 (12 days ago)

Note that PyInstaller and similar also require the packager to determine what data files from each dependency are needed. (using a PyInstaller .spec file)
When new releases of dependencies change their data files, there is no integrated way to know that it needs to be added to the .spec file.
When each package is being tested with its own test suite, it will fail if a mandatory data file was omitted.

The need to maintain the "PyInstaller .spec" for each package's test suite to run makes verifiability a difficult process to set up, and likely becomes a constant source of unnecessary breakages.
Some upstream enhancements in PyInstaller may be possible.

I've only attempted PyInstaller and Nuitka so far.

*Console-less and extensibility*

In order to allow re-use of the story cli , such as wrapping it in another language, or perhaps in a GUI, it needs to be able to operate without a console.
If someone writes a PowerShell script to automate story cli, they dont want console windows popping up and closing.

This has been an ongoing problem with many of the freezers that are simply CPython wrappers, especially relevant on Windows.

An important market segment to reach into is .NET .  There does not appear to be any viable Python 3 in .Net world, not now nor likely in the immediate future.
By freezing with Nuitka, which produces real native PE files on Windows, it is feasible to add .NET support.
There are other freezers like PyOxidizer which would also allow this.

*Static mode & portability*

Specifically with PyInstaller, it has poor support for statically linked Python.  Anaconda now has it working, but the code in PyInstaller for it seems to be very anaconda specific.
Statically linked Python will allow for smaller executables, and better portability.

Note: PyInstaller created apps only work on a single OS, including each distro.  i.e. it does not include libraries that it believes are OS libraries, such as those included in `/usr/lib`.
This means that there is no reliability when it comes to using PyInstaller with system libraries on Linux.
e.g. PyYAML will fall back to using pure Python mode if it was compiled against libyaml when created and the library can not be found on the host, or worse could fail if libyaml is incompatible.
However many other Python libraries which rely on system libraries do not have a pure-Python fallback, and/or that mode isnt being adequately tested downstream or in our CI.

The above also highlights why PyInstaller output appear quite small -- it doesnt include all the system libraries by default.

And Linux distros, and Windows editions, have wildly different system libraries, so using PyInstaller is a constant battle when trying to keep it a portable app.

https://github.com/pyinstaller/pyinstaller/issues/4047 is a common example of the type of problem people encounter.

*Control of the source*

Most of Python development processes and tools rely on `pip`, which either installs whatever is in a wheel, and `setuptools` which at runtime installs whatever it likes.
Patching of the installed packages before PyInstaller is run is not supported.

The most common reason for patching packages is devendoring and removing test suites, however critical fixes are also often carried around by distros as patches,
and typically are applied and released before CVEs are announced.

A recent example of a major problem with vendoring is the urllib3 CVEs of 2019, because urllib3 is vendored by requests, and requests is vendored by pip, etc, etc.
In a frozen app, all of those vendored copies need to be removed/updated in the event of a CVE.

Regarding devendoring, to do this safely the package test suite should be run on the devendored installed version, as often the de-vendored version is incompatible with the package it was vendored in.  Then a patch needs to be created.

*Total cost of ownership*

A serious concern with regards to Nuitka, especially as it is new technology, and building Open Build System as it is also new to Storyscript, is the increase in total cost of ownership.

The current Nuitka build process involves extreme tests invocation in order to validate the platform, and to workaround bugs.

It should be possible to generate the same output on conventional CI in a fraction of the time, without all the validation, especially once a few patches have been submitted and accepted upstream.

This is very likely to be achievable soon on Linux, but may face some difficulties on Windows if using MinGW due to bugs in MinGW gcc which are not going to be fixed quickly.

Once that is done, a simple script would be run in CI, and a subset of the test suite would validate the exe.

The build system on OBS would continue to do complete validation, that could then highlight incompatibilities that the conventional CI wont find, but it would likely become irrelevant as the story QA improves.

The build system on OBS relies heavily on the openSUSE Python package macros, which have in-built support for build "flavours".
Users then choose the flavor with use Debian "alternatives" system (`/etc/alternatives`).

In addition to CPython 2.x and 3.x flavours, it has support for PyPy 3.  The PyPy 3 support has bit-rotted a little due to PyPy 3 being de-supported on openSUSE/SuSE Linux Enterprise (SLE) before its recent resurgence and almost flawless compatibility with CPython, however I am confident that it will be re-enabled soon as I am the person responsible for landing it.

Adding MinGW flavours requires some significant additions to the macros.  I have not discussed these with the other openSUSE Python maintainers yet.
If they are not interesting in maintaining MinGW Python packages, the alternative is for the openSUSE MinGW maintainers to take ownership of them.
The first batch of changes are already done and can be seen at
https://build.opensuse.org/package/rdiff/home:jayvdb:wine-python3:py37:packages/python-rpm-macros?opackage=python-rpm-macros&oproject=devel%3Alanguages%3Apython%3AFactory&rev=9

The Nuitka 'flavor' is more interesting/problematic.  It is quite possible that a distro like openSUSE (community distro) might provide a Nuitka flavour in the future, but it is unlikely to make it into SLE (enterprise) until it can show significant benefits over CPython and PyPy.

Thankfully, the Nuitka flavor requires only tiny changes to the lua macros, so I expect to twist some arms to get the Nuitka flavour merged in, but may need to occasionally fix any breakages myself.

Also the msys & anaconda teams are likely to want to pick some of this up, as fat binary Python libraries aide their mission of allowing multiple apps to co-exist even when their dependencies make them incompatible with the normal Python single-site-packages mechanism.

**Comment by jayvdb** [comment]
Another freezer I haven't looked at yet is https://github.com/beeware/briefcase ; BeeWare has some really good tools, and an active core development team that is a pleasure to collaborate with, so I suspect this is another one worth exploring.

**Comment by jayvdb** [comment]
A quick summary of `briefcase` -- it works by converting Python to Java first, using `voc`.  voc seems a little buggy on Python 3.7, but I havent run the full test suite on Python 3.7 yet.  It is very good on Python 3.5 and seems good on Python 3.6 so far.

Like Nuitka, the `voc` test suite includes running against the entire CPython test suite -- a very encouraging sign that they are solving the problems properly rather than fiddling with each new bug that is reported.  Also like Nuitka, it looks like it can be used to test each component using the test suite of that component, again an indicator of good engineering.  `voc` tests running at https://build.opensuse.org/package/show/home:jayvdb:py-new/python-voc

I've also uploaded a demo project using PyInstaller at https://build.opensuse.org/package/show/home:jayvdb:storyscript:pyinstaller/python-story.  The .spec file there is definately publicly accessible, and includes some comments to explain setting up different ways.  The build log files for the Tumbleweed job (py37) show a constant `ImportError`s , and also package data not being included.  You may need to create an OBS account to see the logs, so I am attaching it here also.
[pyinstaller-story-cli-py37-tumbleweed.log](https://github.com/storyscript/community/files/3690146/pyinstaller-story-cli-py37-tumbleweed.log)

Just a note for anyone else who wants to devote some time to PyInstaller, on my local computer I noticed PyInstaller runs excruciatingly slowly unless it is in a venv.  It seems that it loads into memory the metadata for all installed packages, even though they are not referenced.  Here is a non-debug log showing it going to all strange packages.  [pyinstaller.log](https://github.com/storyscript/community/files/3690164/pyinstaller.log)  So use a venv.  I was able to get much further with PyInstaller by using lots of command line args and a .spec, but it still didnt work on Python 3.6 or 3.7, on Linux or Windows.  As mentioned earlier, testing libraries separately was even more challenging to do en-mass.

**Comment by jayvdb** [comment]
|                                         	| conda/ pipx/ brew/ pynsist/ etc 	| pyi  onedir 	| pyi  onefile 	| Nuitka shared 	| Nuitka static 	| PyOzidizer 	| briefcase 	| cx_Freeze 	|
|-----------------------------------------	|---------------------------------	|-------------	|--------------	|---------------	|---------------	|------------	|-----------	|-----------	|
| story cli dependency support            	| ✔                               	| ✔           	| ✔            	| ✔             	| ❔             	| ✘          	| ❔         	| ✘         	|
| Windows + Linux + macOS support         	| ❔                               	| ✔           	| ✔            	| ✔             	| ❔             	| ✔          	| ✔         	| ✔         	|
| CI for Windows + Linux + macOS          	|                                 	| ✘           	| ✘            	| ✘             	| ✘             	| ✔          	| ✘         	| ✘         	|
| Single file redist                      	| ✔                               	| ✤           	| ✔            	| ✤             	| ✔             	| ✔          	| ✔         	| ✔         	|
| Portable app                            	| ✔                               	| ✔           	| ✔            	| ✔             	| ✔             	| ✔          	| ✔         	| ✔         	|
| Licensing (GPL gdbm)                    	| ✔                               	| ✤           	| ✤            	| ✤             	| ✘             	| ✔          	| ❔         	| ❔         	|
| Licensing (GPL readline)                	| ✔                               	| ✤           	| ✘            	| ✤             	| ✘             	| ✔          	| ❔         	| ❔         	|
| Licensing (LGPL runtime, e.g. libcrypt) 	| ✔                               	| ✤           	| ✤            	| ✤             	| ✤             	| ✔          	| ❔         	| ❔         	|
| Licensing (LGPL chardet)                	| ✔                               	| ✤           	| ✤            	| ✤             	| ✤             	| ✤          	| ❔         	| ❔         	|
| Licensing (PyPI deps)                   	| ✔                               	| ✤           	| ✘            	| ✤             	| ✤             	| ✤          	|           	| ❔         	|
| spafit Linux                            	| ✔                               	| ✔           	| ✔            	| ✔             	| ❔             	| ❔          	| ❔         	| ✘         	|
| Test suite of tool                      	|                                 	| 4           	| 3            	| 5             	| 3             	| 2          	| ✔         	| 1         	|
| Test suite execution on output          	| ✔                               	| ✔           	| ✘            	| ✔             	| ✘             	| ✔          	| 4         	| ✔         	|
| pytest support                          	| ✔                               	| ❔           	| ✘            	| ✔             	| ✘             	| ❔          	| ❔         	| ❔         	|
| Bridge to other platform                	| ✘                               	| ✘           	| ✘            	| C & C++       	| C & C++       	| Rust       	| Java      	| ✘         	|
| Code optimizer                          	| ✘                               	| ✘           	| ✘            	| ✔             	| ✔             	| ❔          	| ❔         	| ✘         	|
| Code smell detector                     	| ✘                               	| ✘           	| ✘            	| ✔             	| ✔             	| ❔          	| ❔         	| ✘         	|

**Comment by jayvdb** [comment]
The first column above groups together solutions which utilise a normal installation of CPython from a vendor that supports it, but use various mechanisms to allow package installations that can not be broken by other package installations.  They need to be installed, and the installed copy uses a normal file based Python code layout.  I havent added columns explicitly for Anaconda, Enthought Canopy, WinPython, and ActivePython , which all provide a non-pip package manager that are an *application* installer rather than a package installer.  Often they also include a GUI for their application package manager.  Of those, I havent experience with Enthought Canopy or WinPython, but the others are well known and can also be considered as part of the first column.

Another set which can be considered part of the first column are Python freeze solutions that strictly use the standardised Python extensible module loading framework.  This includes distributing packages as .egg's, .whl's or provide a custom module format loader, compress them all together and add an installer or .exe header to decompress the components, either as part of installation or at runtime.  `shiv` and `PEX` are an example of this.  `cx_Freeze` might also fit into this category.  However, these often have similar problems to the other tools in the grid, in that they have to guess what to do with `__file__` whenever they encounter it.  I've documented our exposure to this `__file__` problem at https://github.com/storyscript/cli/issues/252#issuecomment-540483233

https://pypi.org/project/egenix-pyrun/ is another solution that I have seen only in passing,  interesting as it is a supported commercial product.  Last release was late 2016, and supports Python 3.7.  https://ep2019.europython.eu/talks/wGmo3yT-pyrun-shipping-the-python-37-runtime-in-just-48mb/

PyInstaller onefile mode especially is not anything like those.  I am singling it out as it looks so easy.  No matter how simple PyInstaller may make it seem, the result is not CPython X.Y.  More importantly, the result is the core of CPython X.Y plus a lot of pyi voodoo to hack around CPython's default behaviour, and adding some functionality to help developers/packagers, such as its own archive format and tools and a custom set of files within the redist.  It is easy enough to create executables without needing to know the internals, but given a large enough user base we will need someone who knows the internals intimately.

Most importantly, PyInstaller is a licensing nightmare because they do not go far enough in building their own CPython, so they bundle up bits and pieces of the PSF distributed CPython and disregard any licensing that the PSF has carefully complied with in their distribution.  Nuitka has the same problem with their static support, however their static support is basic, requiring CPython expertise to get operational and doesnt pretend to be a user-friendly redistributable output.

Nuitka and PyOzidizer are similar to PyInstaller, except have a smaller user bases and lower compatibility, however they provide more features and future potential.  PyOzidizer is exceptional because of its bottom up attention to licensing, and static single exe being its primary purpose.  Nuitka is exceptional because of it is an optimising compiler and ability to interoperate with CPython allowing effortless use of existing test suites, allowing it to be a target runtime for CI, putting it in the same league as CPython and PyPY (and Jython on Python 2.7, but that is largely irrelevant now).

However now is not a good time to adopt PyInstaller, if possible.  With the maturing of Nuitka and PyOxidizer and similar, users of PyInstaller will be transitioning away.
PyInstaller primarily reduces the size and complexity of the components to be redistributed.  There is little motivation for PyInstaller to be considered a testing target by all parts of the Python community.
On the other hand, users and package maintainers are interested in supporting Nuitka and PyOxidizer, because they offer the users and developers new functionality, such as static analysis of the codebase identifying code which smells or can be optimised.

Supporting a PyInstaller runtime on three platforms will be an undesirable drain on resources.  If we expect users to use our PyInstaller version, we should also be using that runtime for every day development activity.  This is possible with Nuitka, and PyOzidixer to a less extent.  OTOH, it is unlikely that our developers will prefer the PyInstaller version over CPython+pip, which means the users of our PyInstaller version suffer.

There are also Python to JavaScript transpilers which have not be assessed, based on prior experience that they are not viable for modern CPython code, and the fact that the result is still reliant on a JavaScript so extra steps are needed.

Summary
-------

pipx remains the most viable low effort option for our rapidly evolving codebases, except that its OS support is unknown.  The investment required to have pipx working and maintained by vendors is expected to be low upfront, and negligible after inclusion in vendor package collections.  It is very likely to be the option with the least TCO for the medium term.  pipx can be used in each repository to install and test the project, catching any problems early.

PyInstaller onefile is likely viable if Storyscript does not use PyInstaller for Linux due to libreadline (unless a commercial license of readline is available? unlikely).  The lack of CI for OSX is a cause for concern, as is the depth of the test suite.  PyInstaller should not be considered if CD of the installer is required.  At present, the generation of the installer must be done by a human and licensing rechecked for each release.  Alternatively a tool needs to be found or built to determine the licenses of all dependencies and include all relevant files into the redist.  Manual test plans are needed for the generated redist, however a small investment of effort improving PyInstaller may lead to the possibility of automation of pytest on the PyInstaller onefile / onedir output (
https://github.com/pyinstaller/pyinstaller/issues/2009).

Nuitka static is too immature, lacks the benefit of being easily testable, and appears to not be an important feature in the short term for its developers.

PyOzidizer is ideal, except for a few bugs.  Most importantly its main developer appears to understand the importance of licensing.

Nuitka shared libs is mature enough to use for storyscript, avoids most of the runtime licensing issues, but requires an installer.  High quality installers for each platform requires per-platform solutions, such as `pynsist` and `dmgbuild` for Windows and macOS respectively.  There is also the possibility of adding https://github.com/JonathonReinhart/staticx on Linux, but that re-introduces the licensing issues.

PyInstaller onedir has some minor advantages over Nuitka shared libs, mostly due to maturty, but it comes at the cost of lack of testability and PyInstaller being a dead-end solution with little growth opportunity.

Recommendations
---------------
If we need a solution right now:

Use pipx in the short term as the recommended solution. Do the necessary work to get it passing tests on our supported OS/distros, including packaging needed to add it to https://github.com/msys2/MINGW-packages , brew, Ubuntu and Fedora.  Possibly also conda.

pipx community is vibrant enough and motivated to assist with support, and any issues in our dependencies on pipx can be raised in those dependencies and expect reasonable support (as opposed to trying to get free support for their package combined with any shrinkwrapping tool).

Other work to improve options for the future:

1. Try building the cli on briefcase, and identify any missing features, creating issues upstream where needed.
2. Patch bom-open to use a non-LGPL alternative to chardet.
3. Add CPython libedit builds from https://github.com/indygreg/python-build-standalone/ in our CI for hub sdk, storyscript and cli , ensuring that development can not introduce dependencies on CPython components that use libreadline or gdbm.
2. Add Nuitka shared lib jobs to our CI for hub sdk, storyscript and cli, testing compatibility
4. Forcefully remove `chardet` from the CI environments before running the test suites.  requests uses chardet for `.text` attribute, while storyscript uses `requests` for the `.json` attribute.  Our CI should pass with `chardet` replaced with a drop-in replacement or a dummy package which only has a `__version__` (requests will emit a warning, but shouldnt break unless we use `.text`)
5. Help fix the `__file__` bugs in our dependencies and possibly assist PyOzidixer for these problems, e.g. adding QA for our dependencies.
6. Help PyOzidixer solve the PyPI package licensing problem, as this will set an example for the other tools to follow, and possibly also provide tools the others can re-use.  The other tools are unlikely to lead in this area.

**Comment by jayvdb** [comment]
Note there is a specific problem in pipx with click, https://github.com/pipxproject/pipx/issues/233 , and issues raised about Windows https://github.com/pipxproject/pipx/issues/81 , but it passes all tests on Windows except for one which relates to symlinks.  The click one could be a serious road block, but the specifics of when it occurs is a bit unclear atm.

**Comment by williammartin** [comment]
@jayvdb wow, very extensive investigation of the existing tooling - thank you so much. I have to say, I'm having a bit of a hard time parsing the vast quantity of information here even with your explanations and summaries. 

Firstly, now that you have an idea for going forward, is it possible to split this into incremental pieces of work such that they might be independently understood by anyone on the team without having all the knowledge in your head?

Secondly, I have a bit of a concern that this is frankly an overwhelming amount of nonsense in the python ecosystem to ship binaries as we'd like them. However, it's possible that with the quantity of stuff in here, I might be misunderstanding the complexity. I'm particularly concerned that even if we resolve this now, the maintenance burden of staying on top of this stuff, particularly with you as a SPOF. If this landed on my plate, I'm not sure that I could proceed. Do you have any thoughts on this?

**Comment by jayvdb** [comment]
@williammartin , the bottom of the post two above has some fairly concrete and discrete actionables : i.e. mostly issues that need to be solved , especially in the CLI, but also dependency projects - ours (sdk and storyscript repo) and dependencies maintained by others.  There are some terms there that may not be immediately groked by everyone, some may need some googling, and I'll work on adding more hyperlinks to it so it is clearer.  Let me know if anything there isnt making sense, wrt the task I am recommending to be done.

Why those task need to be done is more about strategy, and everyone is likely to have their own ideas about strategy.  Specific queries are more useful than me attempting to dump my brain here in the hope it a) is more understandable and b) it will be convincing.  There are other ways to tackle this, and people who have the time to invest in it should dig in a bit and come up with their own strategy/recommendations, and we can do both strategies at the same time, and the result will be even better.

I also have a blog post draft coming up which will help shed light on the bigger picture, but it wont assist with the nitty-gritty details.  In short, the Python world is on the cusp of having some well engineered solutions to this problem.

Beyond those actions, there is a lot of work to be done on improving the test suite of our CLI, in order to be able to validate any CLI we give to our users.  QA is more important *if* we are going to give users a CLI which we do not use internally on a regular basis, and automated QA is necessary given our capacity.  It seems like a lot of our staff are using macOS, so we might want to lead with a fat binary for that platform, even if it isnt the biggest userbase.  macOS also doesnt have GNU readline, which eliminates the largest of the two main legal issues.

You are very close to the mark regarding binaries on Python.  It is one of the worst possible languages for non-technical user CLIs, if the audience are deemed unable to handle installing Python from Python Software Foundation plus some other CLI commands to get it installed.  Python has, until recently, one of the worst dependency management tools of the major languages, and the new tools with real dependency solvers are primarily aimed at development/isolated/virtual environments, i.e. involve lock files, not complex user environments.  Distros have a constant headache trying to get a tiny subset of the Python libraries co-existing in the same namespace.  Other languages have constant buildbots running on all libraries as they are uploaded.  PyPI cant even tell you what the dependencies are of each library, let alone assemble those dependencies into a VM in order to run a test suite, and even if it could, there is no reliable way for a buildbot to programmatically determine how to run the test suite.

PyInstaller and similar have been a constant battle with trying to work with a rapidly evolving Python core, and workarounds for all the dodgy stuff that Python library developers do which was only ever tested on CPython, and often not rigorously tested.  Getting a high quality installer from this requires a concerted effort, at each release, and I expect our rapid development, constant release process and QA problems would mean our installers through this type of technology would be replacing one set of user installation problems which we could blame on upstream (i.e. PSF or libraries) for a new set of problems, but these new ones would be bugs that we need to own.  And it seems PyInstaller has little interest in the legalities, having "wontfixed" a few related issues, and their developer team seems woefully thinly stretched.  There are some PyInstaller users who have created their own solutions for some of the legal issues, and we might be able to get some traction on it, but my guess is that we would be mostly on our own with "owning" the legal side of the binaries produced with this technology.  I can not see PyInstaller accepting that they need to compound their troubles by helping packagers exclude GNU readline, or LGPL system libraries or LGPL Python libraries.  That would be admitting that there is a legal problem, and would scare a large chunk of their existing userbase.

Nuitka and PyOxidizer are a fresh approach, being build tools that are more like compilers than wrappers.  They emit lots of warnings and even error when the input will cause the output to be unusable.  That approach is needed to subdue a language which is built around "duck-typing", in order to get a reliable result.  Nuitka is more advanced but they havent put much effort into legal issues, and it doesnt create its own CPython and static builds are hard when PSF doesnt provide suitable static artifacts (Anaconda does distribute static artifacts, and Nuitka has limited/fragile support for Anaconda).  However PyOxidizer is more principled and provides pre-built CPython that their team has legally vetted and has all the ingredients for static linking, and emits JSON metadata regarding licensing of all of those ingredients.

A concerted effort to solve the few remaining issues wrt Nuitka and PyOxidizer are likely to result in a solution where our ongoing TCO is quite low.  Both projects want the solutions that we need.  Once merged, there is little reason to believe they would bitrot.  I am more confident in PyOxidizer's institutional support, but I foresee Nuitka having a huge adoption by the Python community, and it has PSF and Google support, albeit limited due to it not being their project.  PyInstaller will never have that adoption, because what it generates is something that is, by definition, lower quality than using CPython.  Developers and most technical users will prefer a CPython install over a PyInstaller install, unless they are not able to install & manage CPython on their machine (i.e. the non-technical user).

As a result, I am reverting to investigating a non-fat-binary approach of `pipx` as a quick and dirty approach which should buy us some time, primarily solving the pip problems and the result is very close to what our development team would ordinarily be using day to day.  If we can get `pipx` working reliably for a significant cohort of our current target market, we can look into @judepereira 's suggestion of wrapping a CPython+`pipx`+"story" into a single installer to expand the target market to include people who would have trouble installing the correct CPython, or would balk/hesitate at a two or three stage install (CPython , then pipx, then story).

**Comment by jayvdb** [comment]
Quick analysis of pipx like tools, only eval on Linux atm
- `vam install story` : creates venv `/home/jayvdb/.vam/story/` and script `/home/jayvdb/.local/bin/story`
- `vpip install story` : works after some patches - one more PR to go to create `~/.local/bin/story`. author uses Windows.
- `vex -m story-vex pip install story` : only creates and installs (already on Fedora: https://repology.org/project/vex/versions)
- `venvs create -i story story` creates `/home/jayvdb/.local/share/virtualenvs/story` with story pre-installed. no bin installed in `~/.local/bin/`, and that seems out of scope.
- `fades --check-updates -d story --exec story` - re-uses virtualenv, and updates packages, but doesnt create `~/.local/bin/story` -- seems within their scope to create `~/.local/bin/story`
- `pew new -i story story-pew-venv` creates `/home/jayvdb/.local/share/virtualenvs/story-pew-venv/bin/python3` and launches subshell "Launching subshell in virtual environment. Type 'exit' or 'Ctrl+D' to return.", and has the option of installing a new Python runtime using `pythonz`
- `pipcmd install -p python3 story` installs into `/home/jayvdb/.pipcmd/env/story/`;  no bin installed in `~/.local/bin/`

**Comment by jayvdb** [comment]
Another alternative which I havent put too much attention on is to do something similar to `get-pip.py` , but include an entire venv contents into the script.  https://www.pylucid.org/en/about/ is an implementation of this concept.  The voodoo behind it is https://github.com/jedie/bootstrap_env .  I've seen a few others also, but wasnt paying much attention to that approach.  This approach would remove the extra step of installing `pipx`/similar, but would result in a complete viable venv that the user can then continue to maintain themselves.  PEX & shiv may have something like this, where the included bundle can be exported to a new venv.

note to self mostly, https://github.com/kiwi0fruit/pyappshare is a nice conda based builder, with links to other conda tools at the end of the README.

**Comment by jayvdb** [comment]
I have created a hook plugin system for PyOxidizer, to allow adding of import hacks like adding `__file__`, etc, and now have a working version of the cli where the test suite of hub, storyscript and cli all pass, and almost all tests of almost all of the dependencies pass, when the test suite is run against the story cli binary , if it has some extra test dependencies added to the binary.

There are some concerning test failures, but none that appear to cause us any immediate problems.  Notes at https://github.com/indygreg/PyOxidizer/issues/171  See especially `requests` and `peewee`, each with one relevant test failure that I havent diagnosed yet.  I havent enable `peewee` built extensions yet, which may be why the test fails.

The process of running all tests isnt as easily maintainable as it was with Nuitka, as I do not yet have PyOxidizer set up on Open Build System openSUSE, so I can not leaverage their test invocation logic to run against a large set of packages that are being updated regularly.  But it is definitely feasible to be running the test suites of hub, storyscript and cli, and other pytest-5.x-compatible projects, but fetching the correct version of the tests for each dependency will require a bit of effort.  I hope to have the test-ability of deps become a PyOxidizer feature; c.f. https://github.com/indygreg/PyOxidizer/issues/141

I have also run the CPython test suite again PyOxidizer and roughly categorised the errors. https://github.com/indygreg/PyOxidizer/issues/152 As a result I am concerned about the cli custom tracking/tracing code as it does forking, but that needs to be enhanced anyway due to other problems.

I'll get my hook plugin system published tomorrow, and CI builds of the CLI also up, and then we can evaluate that option.
